source/retryable-writes/tests/README.rst

```diff
@@ -395,6 +395,50 @@ and sharded clusters.
 
    9. Disable the failpoint.
 
+#. Test that drivers return the original error after encountering an error with
+   label `NoWritesPerformed`. This test MUST be implemented by any driver that
+   implements the Command Monitoring specification.
+
+   1. Create a client with ``retryableWrites=true``. If testing against a
+      sharded deployment, the test runner MUST ensure that the client connects
+      to only a single mongos host.
+
+   2. Configure a fail point with error code ``10107`` (NotWritablePrimary)::
+
+         db.adminCommand({
+                configureFailPoint: "failCommand",
+                mode: {times: 1},
+                data: {
+                        errorCode: 10107,
+                        errorLabels: ["RetryableWriteError"],
+                        failCommands: ["insert"],
+                },
+        });
+
+   3. Via the command monitoring ``CommandFailedEvent``, configure a fail point
```

Comment:
I don't think this is possible in all drivers. I believe some drivers use async publishing on command events where the retry can run before the handler. Is there any other way to configure 2 different errors on subsequent attempts?

Reply:
A separate can of worms (albeit unlikely) is that running database commands from _within_ a APM handler seems a possible way to enter a feedback loop. If the `failCommand` command itself were to fail, we might re-enter the existing CommandFailedEvent handler and issue a second `failCommand`. That said, this would be pretty easy to work around in the test implementation (e.g. check that the failed command was `insert` and/or had the previously expected `10107` error code).

Reply:
IIUC there is no way to sequentially issue fail points; the server will only register the most recent configuration. I can only think of two alternatives to using command monitor:

1. Explore using a mock dialer
2. Create a blocking server ticket to register multiple fail points

What are your thoughts?

Reply:
I think we should do both 1) and 2). That is, drivers that can use the CommandFailedEvent event option should be encouraged to do so (perhaps mention to filter based on the first failed _insert_ command only to deal with the problem Jeremy mentioned. Then we can also open a new server ticket to request the ability to set multiple different errors in failCommand. Something like:
```js
db.adminCommand({
    configureFailPoint: "failCommand",
    mode: {times: 2},
    data: {
        failCommands: ["insert"],
        errors: [
            {
                errorCode: 10107,
                errorLabels: ["RetryableWriteError"],
            }, {
                errorCode: 9001,
                errorLabels: ["RetryableWriteError", "NoWritesPerformed],
            }
        ]
    }
);
```

Reply:
How prescriptive should the mocking instructions be? Should I pair with a driver that would require this implementation to make sure we have a comprehensive prose test for this case? Or is it enough to say something like "Drivers MAY use a mock connection class for testing the NoWritesPerformed error handling"?

Reply:
I don't think it needs to be prescriptive since it will vary based on the driver. We can say something like "drivers that are unable to set failCommand after the CommandFailedEvent SHOULD use mocking or write a unittest to cover the same sequence of events".

Reply:
Great, thank you!